### ISSUE TITLE
FO4Edit: Updated wbCTDA, wbCTDAFunctions, and TCTDAFunctionParamType

### ISSUE THREAD

**Issue by fireundubh** [TES5Edit/TES5Edit#212]
Headings are links to pastes.

TODO: Entry Points/EFPTs. Gibbed has a parser that can possibly dump these from Fallout4.exe.

### [wbCTDA Parameter #1 and Parameter #2](http://hastebin.com/raw/aminuhomol)

* Tested only PERK group
* No found-expected errors in PERK group
* Sorted some long signature arrays alphabetically to facilitate testing
* Updated: `Nov 22, 2015 @ 3:03 AM`

### [wbCTDAFunctions](http://hastebin.com/raw/cowayuboje)

* Added 806 functions based on [Gibbed's functions dump](http://mod.gib.me/fallout4/functions.html)
* No more functions left to be added
* TODO: Update SKSE functions to F4SE when ready
* Updated: `Nov 22, 2015 @ 2:58 AM`

### [TCTDAFunctionParamType](http://hastebin.com/raw/polurazoja)

* Sorted alphabetically
* Added ptString to TCTDAFunctionParamType
* Added ParamType4 through ParamType10 to TCTDAFunction
* TODO: Implement ptString somewhere?

### [wbCTDA Parameter #1 and Parameter #2](http://hastebin.com/raw/uyixazayub)

* Referenceable Object and Damage Type in Parameter 1, for some reason, can't coexist. Switching them around produces errors suggesting they need to be switched around.
* There is now unused data in some CTDAs.
* This is the closest I could get to a proper order.
* Updated: `Nov 22, 2015 @ 4:48 PM`

### [wbCTDAFunctions](http://hastebin.com/raw/iwaqolifer)

* Removed non-CTDA functions
* Changed ptString to ptNone wherever appropriate
* Updated: `Nov 22, 2015 @ 4:48 PM` 

### [TCTDAFunctionParamType](http://hastebin.com/raw/vehozopuge)

* Removed ptString per Zilav
* Removed ParamType 4 through ParamType10 per Zilav
* Updated: `Nov 22, 2015 @ 4:48 PM`

**Comment by fireundubh** [comment]
Looks like I mixed up ptObjectReference and ptReferenceableObject. Will correct.

**Comment by zilav** [comment]
10 parameters for CTDA? There is only place for 3 in CTDA subrecord, more that that just doesn't fit. Unless they increased it in size, but I don't remember any 'unused data' warnings for CTDA when decoding records.

**Comment by fireundubh** [comment]
I'm probably referring to the wrong thing. There are functions that have up to 10 arguments.

**Comment by fireundubh** [comment]
I just updated the pastes for wbCTDA Parameters 1 *and* 2, and wbCTDA Functions, which now includes the 64 functions I left out previously for a total of 806 functions.

I believe the wbCTDA parameter orders still need to be tuned, especially whenever we clear unknowns. I also think there are some function arguments that may not be correct. I tried to use the best ones available, but maybe some new parameters need to be implemented. I defaulted all forms to ptReferencableObject. In any case, xEdit should report those as errors if there are any.

ptString needs to be fully implemented as well, but I don't know how or where to do that.

**Comment by zilav** [comment]
Some of those functions (all?) are definitely not for conditions. CTDA functions accept 3 parameters at max. String parameters are stored in CIS1 and CIS2 subrecords, so the data in CTDA is irrelevent and wbUnion should default to unused/unknown byte array. At least this is how it workes in Skyrim iirc.

**Comment by fireundubh** [comment]
So, let's snip the functions with more than 3 parameters.

As for CIS1/CIS2, do string parameters need to be declared in the function definitions? How do you handle them correctly in wbCTDAFunctions?

**Comment by Hugues92** [comment]
There are 2 versions of the functions the "Execute" which was used by ObScript or the console and the "Eval" used by Conditions. The only ones we need in xEdit are the eval versions. Those were limited to 2 parameters, then Skyrim added a third. As far as I checked, it was still 3 for FO4.
(By the way, I received a mail for a comment from Zilav, but it doesn't display here :) )

**Comment by Hugues92** [comment]
Eval functions have a non null value in the field at offset 0x038

**Comment by zilav** [comment]
Ignore string parameters, set those functions as not having parameters ptNone. In this case decider for wbUnion will result in unused byte array and string will be there is CIS1/CIS2, just like in Skyrim.

**Comment by fireundubh** [comment]
I stripped the unnecessary functions, so now I'm reordering the wbCTDA parameters all over again.

> Ignore string parameters, set those functions as not having parameters ptNone.

Okay.

**Comment by fireundubh** [comment]
Is it worth stripping the unnecessary functions? Reordering the parameters is such a pain. I don't think I'm going to be able to do it again. It's like a Rubik's Cube.

Surely there's a better way to implement wbCTDA parameters? Maybe some sort of autodetection or autosort? I'd bet that that this puzzle could be solved algorithmically and far faster than this modify->compile->test->repeat cycle I've been doing for the past 48 hours or so.

**Comment by zilav** [comment]
Doesn't matter much as long as they are not used in CTDA functions array for any parameters. 

What so bad about CTDA? I already described how param types enumeration, union and CTDA array work together. I didn't cycle much when doing Skyrim definitions, and it must be done like once every 3 years. Tedious, but not that bad.

**Comment by fireundubh** [comment]
I'm missing something, and you seem to know how to do this better. I ordered the parameters as correctly as I could, but I hit a roadblock I can't solve.

When I move Referenceable Object above Damage Type, I get an error that suggests Damage Type should be moved above Referenceable Object, and when I do that, I get the opposite error.

And now there's unused data in some CTDAs for some reason.

I posted new links under an Alternatives heading. I excised the unnecessary functions, so these are the sections of the code that correspond to the shortened wbCTDAFunctions definition.

I'll leave ordering the parameters to you now.
